### ISSUE TITLE
Implicit priority in companions of the ancestors of a type

### ISSUE THREAD

**Issue by julienrf** [scala/bug#10427]
Consider the following cocde:

~~~ scala
object Priority {

  class Ev[-A, +B]

  class Foo[A]

  class Bar extends Foo[Quux]

  class Quux

  object Foo {
    implicit val ev: Ev[Any, Boolean] = new Ev
  }

  object Bar {
    implicit val ev: Ev[Any, String] = new Ev
  }

  object Quux {
    implicit val ev: Ev[Any, Int] = new Ev
  }

  // Here we require an implicit parameter of type `Ev[Bar, ?]`.
  // Candidates will be looked up from the companion objects of the
  // types associated with `Bar`: `Bar`, `Foo` and `Quux`
  def usage[X](implicit ev: Ev[Bar, X]): Unit = ()

  usage
}

**Comment by milessabin** [comment]
I agree that it's maybe a little counterintuitive that `Quux.ev` has the same priority as `Bar.ev`, but I think it's pretty clear from the spec why that is. `Bar` extends `Foo[Quux]` so `Bar.ev` has higher priority than `Foo.ev`, but `Bar` _doesn't_ extend `Quux`, so `Bar.ev` and `Quux.ev` have the same priority.

In line with the spec you could reduce the priority of `Quux.ev` relative to `Bar.ev` by having `Bar` also extend `Quux`,
```scala
trait Quux
class Bar extends Foo[Quux] with Quux
```
but most likely that's not going to make much sense in your real context.

How would you revise the overloading rules to get the effect that you want?

**Comment by julienrf** [comment]
Yes, I agree that according to the spec what we observe is expected.

I think it would make sense to reduce the priority of implicits found in companion objects of types that are part of a type which is associated with the type looked up.

**Comment by som-snytt** [comment]
If C is derived from D, then it is also derived from the parts of D? for SIP language, in "overload resolution." Maybe it's time to rename the feature, too. "Cog selection from a box of gears."

**Comment by milessabin** [comment]
@som-snytt yup, that would be a route we could explore, but it's a semantic change rather than a bugfix.
